### PR TITLE
vm auto-registration: refactor out logic of health status updates

### PR DESCRIPTION
**Pull request by yskopets**
**Please provide a description of this PR:**

Follow-up to https://github.com/istio/istio/pull/44712.

Refactors out smaller sub-components out of `pilot/pkg/autoregistration/controller.go`:
* logic of receiving health status updates - `pilot/pkg/autoregistration/internal/health`
* logic of persisting internal state (health status) as part of `WorkloadEntry` resource - `pilot/pkg/autoregistration/internal/state`

This PR is merely a refactoring:
  * it doesn't add any new features
  * all unit and e2e tests remain unchanged

### CODE REVIEW

File: pilot/pkg/autoregistration/internal/health/util.go [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+
+// IsElegibleForHealthStatusUpdates returns true if a given WorkloadEntry
+// is allowed to receive health status updates sent by an Istio Proxy.
+//
+// Consider a workload elegible for health status updates as long as the
+// WorkloadEntryHealthCheckAnnotation is present (no matter what the value is).
+// In case the annotation is present but the value is not "true", the proxy should be allowed
+// to send health status updates, config health condition should be updated accordingly,
+// however reported health status should not come into effect.
+func IsElegibleForHealthStatusUpdates(wle *config.Config) bool {
```
Comment (bleggett) [inline pilot/pkg/autoregistration/internal/health/util.go:31]:
```suggestion
func IsEligibleForHealthStatusUpdates(wle *config.Config) bool {
```

nit - predates your fix.

**Review by bleggett — APPROVED** [review]
LGTM

File: pilot/pkg/autoregistration/internal/health/util.go (hunk 1 above)
Comment (yskopets) [reply pilot/pkg/autoregistration/internal/health/util.go]: done

**Review by bleggett — APPROVED** [review]

File: pilot/pkg/autoregistration/internal/health/controller.go [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+	}
+	c.healthCondition = controllers.NewQueue("healthcheck",
+		controllers.WithMaxAttempts(maxRetries),
+		controllers.WithGenericReconciler(c.updateWorkloadEntryHealth))
+	return c
+}
+
+func (c *Controller) Run(stop <-chan struct{}) {
+	go c.healthCondition.Run(stop)
+	<-stop
```
Comment (hzxuzhonghu) [inline pilot/pkg/autoregistration/internal/health/controller.go:64]: nit:  these are equal to` c.healthCondition.Run(stop)`

**Review by hzxuzhonghu — COMMENTED** [review]
/lgtm

File: pilot/pkg/autoregistration/internal/health/controller.go (hunk 2 above)
Comment (yskopets) [reply pilot/pkg/autoregistration/internal/health/controller.go]: done

**Comment by howardjohn** [comment]
/retest

bookinfo-build_istio  is a bug unrealted to this, WIP to fix it

**Review by hzxuzhonghu — APPROVED** [review]

**Pull request merged**